packets.js

```diff
@@ -127,21 +127,57 @@ runknown.encodingLength = function (data) {
 
 var rtxt = {}
 
-rtxt.encode = function (s, buf, offset) {
-  str.encode(s, buf, offset + 2)
-  buf.writeUInt16BE(str.encode.bytes, offset)
-  rtxt.encode.bytes = str.encode.bytes + 2
+rtxt.encode = function (data, buf, offset) {
+  var oldOffset = offset
+  offset += 2
+
+  if (!data) data = {}
+  var keys = Object.keys(data)
+
+  if (keys.length === 0) {
+    str.encode('', buf, offset)
+    offset += str.encode.bytes
+  }
+
+  keys.forEach(function (key) {
+    var s = key + '=' + data[key]
+    str.encode(s, buf, offset)
+    offset += str.encode.bytes
+  })
+
+  buf.writeUInt16BE(offset - oldOffset - 2, oldOffset)
+  rtxt.encode.bytes = offset - oldOffset
   return buf
 }
 
 rtxt.decode = function (buf, offset) {
-  var s = str.decode(buf, offset + 2)
-  rtxt.decode.bytes = str.decode.bytes + 2
-  return s
+  var data = {}
+  var oldOffset = offset
+  var len = buf.readUInt16BE(offset)
+  var s, i
+
+  offset += 2
+
+  while (offset < oldOffset + len) {
+    s = str.decode(buf, offset)
+    i = s.indexOf('=')
+    offset += str.decode.bytes
+
+    if (i === -1) data[s.toLowerCase()] = true
+    else if (i === 0) continue // ignore: invalid key-length
+    else data[s.substr(0, i).toLowerCase()] = s.substr(i + 1)
```

Comment:
could you use .slice(startOffset, endOffset) instead of substr? that the only one i normally use